src/lib/hooks/useSearchParamsState.ts

```diff
@@ -3,66 +3,95 @@ import {useCallback} from 'react'
 
 import {getEnumMemberFilter, isEnumMember} from '../utils/common/enums'
 
+export type SearchParamsUpdater = (
+  searchParams: URLSearchParams,
+) => URLSearchParams
+
+export type SearchParamsUpdaterFactory<T> = (
+  newValue: T | null,
+) => SearchParamsUpdater
+
+type SearchParamsSetters<T> = ReturnType<typeof useSearchParamsSetters<T>>
+
+const useSearchParamsSetters = <T>(
+  getSearchParamsUpdater: SearchParamsUpdaterFactory<T>,
+) => {
+  const searchParams = useSearchParams()
+  const router = useRouter()
+  const pathname = usePathname()
+
+  const getUpdatedUrl = useCallback(
+    (newValue: T | null) => {
+      const updater = getSearchParamsUpdater(newValue)
+
+      const updatedParams = updater(
+        new URLSearchParams(searchParams.toString()),
+      )
+
+      return `${pathname}?${updatedParams.toString()}`
+    },
+    [pathname, searchParams, getSearchParamsUpdater],
+  )
+
+  const setValue = useCallback(
+    (newValue: T | null) => {
+      router.push(getUpdatedUrl(newValue), {scroll: false})
+    },
+    [router, getUpdatedUrl],
+  )
+
+  return {
+    setValue,
+    getUpdatedUrl,
+    getSearchParamsUpdater,
+  }
+}
+
 type UseSearchParamsStateFn<Value extends string[] | string | number> = {
   (
     key: string,
     defaultValue?: undefined,
-  ): [Value | undefined, (value: Value) => void, (value: Value) => string]
+  ): [Value | undefined, SearchParamsSetters<Value>]
   <DefaultValue extends Value>(
     key: string,
     defaultValue: DefaultValue,
-  ): [Value | DefaultValue, (value: Value) => void, (value: Value) => string]
+  ): [Value | DefaultValue, SearchParamsSetters<Value>]
 }
 
 export const useSearchParamsState: UseSearchParamsStateFn<string> = <
   DefaultValue extends string | undefined,
 >(
   key: string,
   defaultValue: DefaultValue,
-): [
-  string | DefaultValue,
-  (value: string) => void,
-  (value: string) => string,
-] => {
+): [string | DefaultValue, SearchParamsSetters<string>] => {
   const searchParams = useSearchParams()
-  const router = useRouter()
-  const pathname = usePathname()
 
   const value: string | DefaultValue = searchParams.get(key) ?? defaultValue
 
-  const getNewUrl = useCallback(
-    (newValue: string) => {
-      const params = new URLSearchParams(searchParams.toString())
-      params.set(key, newValue)
-
-      return `${pathname}?${params.toString()}`
+  const getSearchParamsUpdater = useCallback(
+    (newValue: string | null) => (params: URLSearchParams) => {
+      if (newValue === null) {
+        params.delete(key)
+      } else {
+        params.set(key, newValue)
+      }
+      return params
```

Comment:
Search params updaters were added to allow composing multiple search params updates into a single `router.push` call - for example when applying multiple filters and resetting pagination at once. They're best used by the `useUpdateSearchParams` hook, which takes an array of updaters, applies them to the current search params and pushes the updated url.

Reply:
Awesome! So far it seems clear to me. Will get back to you if there is anything unclear when implementing.